### PR TITLE
Added step and linear bdry interpolation methods

### PR DESCRIPTION
**Pull request by Pranav-Jain**

### CODE REVIEW

File: src/polyfem/assembler/GenericProblem.hpp [hunk 1]
```diff
@@ -45,6 +45,43 @@ namespace polyfem
 			double form_;
 		};
 
+		class PiecewiseConstantInterpolation : public Interpolation
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.hpp:48]: These are all pretty much the same class with a different eval function. I suggest creating a parent class 
```c++
class PiecewiseInterpolation : public Interpolation
{
		public:
			void init(const json &params) override;

		private:
			std::vector<double> points_;
			std::vector<double> values_;
			int extend_ = -1; //0-constant, 1-extrapolate, 2-repeat, 3-repeat_offset
			double offset_ = 0.0;
}
```
and then inherit from this class and override eval.

File: src/polyfem/assembler/GenericProblem.hpp [hunk 2]
```diff
@@ -45,6 +45,43 @@ namespace polyfem
...
+		class PiecewiseLinearInterpolation : public Interpolation
+		{
+		public:
+			double eval(const double t) const override;
+			void init(const json &params) override;
+
+		private:
+			std::vector<double> points_;
+			std::vector<double> values_;
+			int extend_ = -1; //0-constant, 1-extrapolate, 2-repeat, 3-repeat_offset
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.hpp:70]: Make this an `enum`

File: src/polyfem/assembler/GenericProblem.cpp [hunk 3]
```diff
@@ -80,6 +86,130 @@ namespace polyfem
...
+					extend_ = 0;
+				else if (params["extend"] == "extrapolate")
+					extend_ = 1;
+				else if (params["extend"] == "repeat")
+					extend_ = 2;
+				else if (params["extend"] == "repeat_offset")
+					extend_ = 3;
+			}
+			else
+				extend_ = -1;
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.cpp:210]: Does FEBio define a default? We should use that same one. or else throw and error here.

File: src/polyfem/assembler/GenericProblem.cpp [hunk 4]
```diff
@@ -80,6 +86,130 @@ namespace polyfem
 				form_ = 0;
 		}
 
+		double PiecewiseConstantInterpolation::eval(const double t) const
+		{
+			if (t < points_[0])
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.cpp:91]: I think the extend applies in this case too.

File: src/polyfem/assembler/GenericProblem.cpp [hunk 5]
```diff
@@ -80,6 +86,130 @@ namespace polyfem
...
+				return 0;
+			for (size_t i = 0; i < points_.size()-1; ++i) {
+				if (t >= points_[i] && t < points_[i+1]) {
+					return values_[i];
+				}
+			}
+			if (extend_ == 0)
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.cpp:98]: This is the same between all `PiecewiseInterpolation` classes, right? If so move it into a function in the `PiecewiseInterpolation` parent class and call that one in the child class when needed.

File: src/polyfem/assembler/GenericProblem.cpp [hunk 6]
```diff
@@ -80,6 +86,130 @@ namespace polyfem
...
+				return values_[values_.size()-1];
+			else if (extend_ == 1)
+				return values_[values_.size()-1]*t;
+			else if (extend_ == 2) {
+				for (size_t i = 0; i < points_.size()-1; ++i) {
+					if (t-points_[points_.size()-1] >= points_[i] && t-points_[points_.size()-1] < points_[i+1]) {
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.cpp:104]: For this just make it a (recursive) call to `eval(fmod(t, points[-1]) + points[0])+offset`.

File: src/polyfem/assembler/GenericProblem.hpp [hunk 7]
```diff
@@ -45,6 +45,43 @@ namespace polyfem
...
+			double offset_ = 0.0;
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.hpp:71]: I think `offset_` should just be `values_[-1]`

File: src/polyfem/assembler/GenericProblem.cpp (hunk 3 above)
Comment (zfergus) [reply src/polyfem/assembler/GenericProblem.cpp:210]: Default is constant in FEBio: https://help.febio.org/FebioUser/FEBio_um_3-4-Subsection-3.17.1.html

File: src/polyfem/assembler/GenericProblem.cpp [hunk 8]
```diff
@@ -80,6 +86,93 @@ namespace polyfem
...
+				}
+			}
+
+			extend(t);
+			return 0;
+		}
+
+		double PiecewiseLinearInterpolation::eval(const double t) const
+		{
+			if (t < points_[0])
```
Comment (teseoch) [inline src/polyfem/assembler/GenericProblem.cpp:103]: This should go, right?

File: src/polyfem/assembler/GenericProblem.cpp [hunk 9]
```diff
@@ -80,6 +86,93 @@ namespace polyfem
...
+			
+			extend(t);
+			return 0;
+		}
+
+
+		void PiecewiseInterpolation::init(const json &params)
+		{
+			if (!params["points"].is_array())
+				throw "Points must be an array";
```
Comment (teseoch) [inline src/polyfem/assembler/GenericProblem.cpp:120]: There is a function log and throw, cna you use that?

**Review by teseoch — COMMENTED** [review]
2 small things

File: src/polyfem/assembler/GenericProblem.cpp [hunk 10]
```diff
@@ -80,6 +86,91 @@ namespace polyfem
...
+		double PiecewiseLinearInterpolation::eval(const double t) const
+		{
+			for (size_t i = 0; i < points_.size()-1; ++i) {
+				if (t >= points_[i] && t < points_[i+1]) {
+					double val = (values_[i+1]-values_[i]) * ((t-points_[i])/(points_[i+1]-points_[i])) + values_[i];
+					return val;
+				}
+			}
+			
+			extend(t);
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.cpp:110]: Shouldnt you return this?

File: src/polyfem/assembler/GenericProblem.cpp [hunk 11]
```diff
@@ -80,6 +86,91 @@ namespace polyfem
...
+
+			if (params.contains("extend")) {
+				if (params["extend"] == "constant")
+					ext_ = constant;
+				else if (params["extend"] == "extrapolate")
+					ext_ = extrapolate;
+				else if (params["extend"] == "repeat")
+					ext_ = repeat;
+				else if (params["extend"] == "repeat_offset")
+					ext_ = repeat_offset;
```
Comment (zfergus) [inline src/polyfem/assembler/GenericProblem.cpp:138]: Else throw error?

**Review by zfergus — CHANGES_REQUESTED** [review]

File: src/polyfem/assembler/GenericProblem.cpp (hunk 10 above)
Comment (Pranav-Jain) [reply src/polyfem/assembler/GenericProblem.cpp:110]: Yes. Good catch. Thanks!

**Review by zfergus — APPROVED** [review]

**Pull request merged**